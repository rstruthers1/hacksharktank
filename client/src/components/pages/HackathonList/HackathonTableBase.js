import React, { useState } from 'react';
import { Table, Pagination } from 'react-bootstrap';
import './HackathonTable.css';

const HackathonTableBase = ({ hackathons, itemsPerPage = 5, columns, renderRow }) => {
    const [activePage, setActivePage] = useState(1);
    const [searchTerm, setSearchTerm] = useState('');

    const filteredHackathons = searchTerm
        ? hackathons.filter(hackathon =>
            hackathon.eventName.toLowerCase().includes(searchTerm.toLowerCase())
        )
        : hackathons;

    const indexOfLastItem = activePage * itemsPerPage;
    const indexOfFirstItem = indexOfLastItem - itemsPerPage;
    const currentItems = filteredHackathons.slice(indexOfFirstItem, indexOfLastItem);

    const paginate = (pageNumber) => setActivePage(pageNumber);

    const totalPages = Math.ceil(hackathons.length / itemsPerPage);
    let items = [];
    for (let number = 1; number <= totalPages; number++) {
        items.push(
            <Pagination.Item key={number} active={number === activePage} onClick={() => paginate(number)}>
                {number}
            </Pagination.Item>,
        );
    }

    return (
        <div className="table-container">
            <div className="search-container">
                <input
                    type="text"
                    placeholder="Search by event name..."
                    onChange={(e) => setSearchTerm(e.target.value)}
                />
            </div>
            <div className="table-scrollable">
                <Table bordered hover>
                    <thead>
                    <tr>
                        {columns.map(column => (
                            <th key={column} className={`${column}Column`}>{column}</th>
                        ))}
                    </tr>
                    </thead>
                    <tbody>
                    {currentItems.map((hackathon) => (
                        renderRow(hackathon)
                    ))}
                    </tbody>
                </Table>
            </div>
            <div className="pagination-container">
                <Pagination>
                    <Pagination.First onClick={() => paginate(1)} disabled={activePage === 1} />
                    <Pagination.Prev onClick={() => paginate(activePage - 1)} disabled={activePage === 1} />
                    {items}
                    <Pagination.Next onClick={() => paginate(activePage + 1)} disabled={activePage === totalPages} />
                    <Pagination.Last onClick={() => paginate(totalPages)} disabled={activePage === totalPages} />
                </Pagination>
            </div>
        </div>
    );
};

export default HackathonTableBase;