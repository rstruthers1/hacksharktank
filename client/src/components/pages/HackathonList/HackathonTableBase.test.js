import React from 'react';
import { render, screen } from '@testing-library/react';
import HackathonTableBase from './HackathonTableBase';
import renderer from "react-test-renderer";
import '@testing-library/jest-dom';

describe('HackathonTableBase', () => {
    const mockHackathons = [
        { id: 1, eventName: 'Test Hackathon 1', description: 'Description 1', startDate: '2022-01-01', endDate: '2022-01-02' },
        { id: 2, eventName: 'Test Hackathon 2', description: 'Description 2', startDate: '2022-02-01', endDate: '2022-02-02' },
    ];

    const mockColumns = ['ID', 'Event Name', 'Description', 'Start Date', 'End Date'];

    const mockRenderRow = (hackathon) => (
        <tr key={hackathon.id}>
            <td>{hackathon.id}</td>
            <td>{hackathon.eventName}</td>
            <td>{hackathon.description}</td>
            <td>{hackathon.startDate}</td>
            <td>{hackathon.endDate}</td>
        </tr>
    );

    it('should render correctly', () => {
        const tree = renderer.create(<HackathonTableBase hackathons={mockHackathons} columns={mockColumns} renderRow={mockRenderRow} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        render(<HackathonTableBase hackathons={mockHackathons} columns={mockColumns} renderRow={mockRenderRow} />);
    });

    it('renders the correct number of rows', () => {
        render(<HackathonTableBase hackathons={mockHackathons} columns={mockColumns} renderRow={mockRenderRow} />);
        const rows = screen.getAllByRole('row');
        // Subtract 1 for the header row
        expect(rows.length - 1).toEqual(mockHackathons.length);
    });

    it('renders the correct data in the rows', () => {
        render(<HackathonTableBase hackathons={mockHackathons} columns={mockColumns} renderRow={mockRenderRow} />);
        const firstRow = screen.getByText('Test Hackathon 1');
        expect(firstRow).toBeInTheDocument();
    });
});