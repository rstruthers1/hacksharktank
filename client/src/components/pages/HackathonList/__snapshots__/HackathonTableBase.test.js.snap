// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackathonTableBase should render correctly 1`] = `
<div
  className="table-container"
>
  <div
    className="search-container"
  >
    <input
      onChange={[Function]}
      placeholder="Search by event name..."
      type="text"
    />
  </div>
  <div
    className="table-scrollable"
  >
    <table
      className="table table-bordered table-hover"
    >
      <thead>
        <tr>
          <th
            className="IDColumn"
          >
            ID
          </th>
          <th
            className="Event NameColumn"
          >
            Event Name
          </th>
          <th
            className="DescriptionColumn"
          >
            Description
          </th>
          <th
            className="Start DateColumn"
          >
            Start Date
          </th>
          <th
            className="End DateColumn"
          >
            End Date
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            Test Hackathon 1
          </td>
          <td>
            Description 1
          </td>
          <td>
            2022-01-01
          </td>
          <td>
            2022-01-02
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            Test Hackathon 2
          </td>
          <td>
            Description 2
          </td>
          <td>
            2022-02-01
          </td>
          <td>
            2022-02-02
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="pagination-container"
  >
    <ul
      className="pagination"
    >
      <li
        className="page-item disabled"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            «
          </span>
          <span
            className="visually-hidden"
          >
            First
          </span>
        </span>
      </li>
      <li
        className="page-item disabled"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            ‹
          </span>
          <span
            className="visually-hidden"
          >
            Previous
          </span>
        </span>
      </li>
      <li
        className="page-item active"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          1
          <span
            className="visually-hidden"
          >
            (current)
          </span>
        </span>
      </li>
      <li
        className="page-item disabled"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            ›
          </span>
          <span
            className="visually-hidden"
          >
            Next
          </span>
        </span>
      </li>
      <li
        className="page-item disabled"
      >
        <span
          className="page-link"
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            »
          </span>
          <span
            className="visually-hidden"
          >
            Last
          </span>
        </span>
      </li>
    </ul>
  </div>
</div>
`;
